pkg/acsengine/engine.go

```diff
@@ -83,6 +83,37 @@ var kubernetesTemplateFiles = []string{kubernetesBaseFile, kubernetesAgentResour
 var swarmTemplateFiles = []string{swarmBaseFile, swarmAgentResourcesVMAS, swarmAgentVars, swarmAgentResourcesVMSS, swarmAgentResourcesClassic, swarmBaseFile, swarmMasterResources, swarmMasterVars, swarmWinAgentResourcesVMAS, swarmWinAgentResourcesVMSS, windowsParams}
 var swarmModeTemplateFiles = []string{swarmBaseFile, swarmAgentResourcesVMAS, swarmAgentVars, swarmAgentResourcesVMSS, swarmAgentResourcesClassic, swarmBaseFile, swarmMasterResources, swarmMasterVars, swarmWinAgentResourcesVMAS, swarmWinAgentResourcesVMSS}
 
+/**
+ ServicePrincipalClientSecret could be either a plain text, or a reference to a secret in a keyvault.
+ The latter has the following syntax in the api model file:
+
+ "servicePrincipalClientSecret": "/subscriptions/<SUB_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.KeyVault/vaults/<KV_NAME>/secrets/<NAME>"
+
+ This will generate a reference block in the parameters file:
+
+ "reference": {
+   "keyVault": {
+     "id": "/subscriptions/<SUB_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.KeyVault/vaults/<KV_NAME>"
+   },
+   "secretName": "<NAME>"
```

Comment:
arm also specifies using SecretVersion as well so that would look like
"servicePrincipalClientSecret": "/subscriptions/<SUB_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.KeyVault/vaults/<KV_NAME>/secrets/<NAME>/<VERSION>"
outputs like  
"reference": {  
   "keyVault": {  
     "id": "/subscriptions/<SUB_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.KeyVault/vaults/<KV_NAME>" 
   },  
   "secretName": "<NAME>",
   "secretVersion": "<VERSION>"
}
If we could allow that format as well it would allow a user to choose to pin a version of a secret.